### DISCUSSION TITLE
Support for `gpt4all`

### DISCUSSION THREAD

**Discussion by SeanMcTex** [aaif-goose/goose#64] [Ideas]
[gpt4all](https://github.com/nomic-ai/gpt4all) makes it pretty easy to download and run LLMs on your own machine. It also has an OpenAI-compatible API.

It would be great to be able to point Goose to a local LLM this way, thereby saving costs and mitigating concerns about what's getting sent to one's LLM provider.

**Comment by michaelneale** [comment]
@SeanMcTex yeah - can currently use ollama for that (which is really openai API to it - and it works ok as long as can run a suitable model). Can gpt4all run large enough models to be worthwhile? (you need fairly hefty ones with some local GPU support to be worth it for goose from my experience). I have used llamafile and llama.cpp as well (antthing that offers a openai like http interface, and has to have tool calling). 

Would ollama be interesting? 

https://github.com/square/exchange/blob/main/src/exchange/providers/ollama.py

**Reply by SeanMcTex** [reply]
Hey @michaelneale ! Thanks for the response. I spun up Ollama and, after a bit of banging around, got it working with Goose. Alas, while it seems a capable model when I'm chatting casually with it, the responses that Llama 3.1 provides Goose are almost comically bad compared to what OpenAI provides. I'll try with one of the versions of Llama with more parameters when I have time to download one of those bigger ones.

And Ollama seems nicer than GPT4All in general, so I'm going to switch to using it for my local LLM needs. So consider this question answered!

**Reply by michaelneale** [reply]
great to hear @SeanMcTex  - I found some ok results with https://ollama.com/library/qwen2 and in particular (but slow on my machine): https://ollama.com/library/qwen2:72b shows promise, so please keep experimenting. Local LLM will get there one day, so worth being ready for it IMO. 

Also, llama3.1 - even if you use the largest one, doesn't do tool calling so it isn't great for goose, but it is possible to use other models to complement things like this: https://github.com/square/goose/pull/68 - in this case it is the new o1 model, which also doesn't do tool calling, but the other traditional models do that, and uses this for more expertise/consulting. That could work here as well, use a mixture of models to get better results.

**Reply by SeanMcTex** [reply]
Oh, very interesting! Quen's super-slow here as well, and the results still seemed less useful than those OpenAI is giving. But I'm sure all of these models will continue to become better and more efficient over time.

The multimodel approach is an interesting one, for sure! Will continue to experiment and watch how these tools develop with interest. Thanks very much!

**Reply by michaelneale** [reply]
@SeanMcTex yes I can imagine todays frontier models are next years run on laptop models!

**Reply by wwilson83** [reply]
@SeanMcTex Any chance you can share a guide or configuration changes you made for connecting to your local LLM instance? I'm currently using LM Studio but a switch to ollama wouldnt be hard.

**Reply by SeanMcTex** [reply]
@wwilson83 Here are a couple of configs I've tried for Goose, pointing to ollama. Hope that's some help!

```
ollama:
  provider: ollama
  processor: llama3.1:70b
  accelerator: llama3.1:70b
  moderator: passive
  toolkits:
  - name: developer
    requires: {}
qwen:
  provider: ollama
  processor: qwen2:72b
  accelerator: qwen2:72b
  moderator: passive
  toolkits:
  - name: developer
    requires: {}
```

**Reply by wwilson83** [reply]
@SeanMcTex Thank you.
